### PR TITLE
8312428: PKCS11 tests fail with NSS 3.91

### PR DESCRIPTION
**Pull request by valeriepeng**
Starting NSS v3.91, SHA-3 support is added for MessageDigest but not for PSS Signature. This breaks existing test assumptions made by PSS regression tests. In addition, the NSS SHA-3 message digests do not support cloning which causes the failure of TestCloning.java.

This PR adds a PSSUtil.java class which provides utility method for detecting/guessing whether a digest algorithm is valid for PSS signature or not.

The changes are verified with NSS v3.46, v3.57 and v3.91 (on local Linux machine).

Thanks in advance for review~

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8312428](https://bugs.openjdk.org/browse/JDK-8312428): PKCS11 tests fail with NSS 3.91 (**Bug** - P3)

### Reviewers
 * [Sibabrata Sahoo](https://openjdk.org/census#ssahoo) (@sisahoo - Committer) ⚠️ Review applies to [ca1a3b92](https://git.openjdk.org/jdk/pull/15217/files/ca1a3b9203f5147b35449dedb9d7b1ff39d3fad9)
 * [Rajan Halade](https://openjdk.org/census#rhalade) (@rhalade - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15217/head:pull/15217` \
`$ git checkout pull/15217`

Update a local copy of the PR: \
`$ git checkout pull/15217` \
`$ git pull https://git.openjdk.org/jdk.git pull/15217/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15217`

View PR using the GUI difftool: \
`$ git pr show -t 15217`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15217.diff">https://git.openjdk.org/jdk/pull/15217.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15217#issuecomment-1672393474)

### CODE REVIEW

**Review by sisahoo — APPROVED** [review]

File: test/jdk/sun/security/pkcs11/PSSUtil.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+           "SHA3-224", "SHA3-256", "SHA3-384" , "SHA3-512",
+    };
+
+    public static enum AlgoSupport {
+        NO, MAYBE, YES
+    };
+
+    public static boolean isSignatureSupported(Provider p) {
+        try {
+            Signature.getInstance("RSASSA-PSS", p);
```
[the remark this answers is not in the document]
Comment (valeriepeng) [reply test/jdk/sun/security/pkcs11/PSSUtil.java:45]: Sure

File: test/jdk/sun/security/pkcs11/Signature/SignatureTestPSS.java [hunk 2]
```diff
@@ -62,82 +62,57 @@ public static void main(String[] args) throws Exception {
 
     @Override
     public void main(Provider p) throws Exception {
-        Signature sig;
-        try {
-            sig = Signature.getInstance(SIGALG, p);
-        } catch (NoSuchAlgorithmException e) {
+        if (!PSSUtil.isSignatureSupported(p)) {
             System.out.println("Skip testing RSASSA-PSS" +
```
Comment (rhalade) [inline test/jdk/sun/security/pkcs11/Signature/SignatureTestPSS.java:66]: throw SkippedException here.

File: test/jdk/sun/security/pkcs11/Signature/SignatureTestPSS.java [hunk 3]
```diff
@@ -62,82 +62,57 @@ public static void main(String[] args) throws Exception {
...
+            PrivateKey privKey = kp.getPrivate();
+            PublicKey pubKey = kp.getPublic();
+            for (String hash : DIGESTS) {
+                for (String mgfHash : DIGESTS) {
+                    System.out.println("    [Hash  = " + hash +
+                            ", MGF1 Hash = " + mgfHash + "]");
+                    PSSUtil.AlgoSupport s =
+                            PSSUtil.isHashSupported(p, hash, mgfHash);
+                    if (s == PSSUtil.AlgoSupport.NO) {
+                        System.out.println("    => Skip; no support");
```
Comment (rhalade) [inline test/jdk/sun/security/pkcs11/Signature/SignatureTestPSS.java:83]: Similar request here to mark test as skipped if all algorithms are not supported.

File: test/jdk/sun/security/pkcs11/PSSUtil.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+import java.security.*;
+import java.security.interfaces.*;
```
Comment (rhalade) [inline test/jdk/sun/security/pkcs11/PSSUtil.java:24]: Please remove unused imports

File: test/jdk/sun/security/pkcs11/PSSUtil.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import java.security.spec.*;
+
+public class PSSUtil {
+
+    /**
+     * ALGORITHM name, fixed as RSA for PKCS11
+     */
+    private static final String KEYALG = "RSA";
+    private static final String SIGALG = "RSASSA-PSS";
+    private static final String[] DIGESTS = {
```
Comment (rhalade) [inline test/jdk/sun/security/pkcs11/PSSUtil.java:34]: is this constant used anywhere or supposed to be used?

File: test/jdk/sun/security/pkcs11/PSSUtil.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            return false;
+        }
+    }
+
+    public static AlgoSupport isHashSupported(Provider p, String... hashAlgs) {
+
+        AlgoSupport status = AlgoSupport.YES;
+        for (String h : hashAlgs) {
+            String sigAlg = (h.startsWith("SHA3-")?
+                    h : h.replace("-","")) + "withRSASSA-PSS";
```
Comment (rhalade) [inline test/jdk/sun/security/pkcs11/PSSUtil.java:59]:
```suggestion
            String sigAlg = (h.startsWith("SHA3-") ?
                    h : h.replace("-", "")) + "withRSASSA-PSS";
```

**Comment by rhalade** [comment]
Can you please also update https://github.com/openjdk/jdk/blob/master/test/jdk/sun/security/pkcs11/PKCS11Test.java#L1002 to use 3.91?

File: test/jdk/sun/security/pkcs11/MessageDigest/TestCloning.java [hunk 7]
```diff
@@ -23,7 +23,7 @@
 
 /*
  * @test
- * @bug 6414899 8242332
+ * @bug 6414899 8242332 8312428
```
[the remark this answers is not in the document]
Comment (valeriepeng) [reply test/jdk/sun/security/pkcs11/MessageDigest/TestCloning.java:26]: Ok.

File: test/jdk/sun/security/pkcs11/MessageDigest/TestCloning.java [hunk 8]
```diff
@@ -58,9 +58,16 @@ public void main(Provider p) throws Exception {
...
         for (String alg : ALGS) {
-            System.out.println("Testing " + alg);
+            System.out.println("Digest algo: " + alg);
             MessageDigest md = MessageDigest.getInstance(alg, p);
-            md = testCloning(md, p);
+            try {
+                md = testCloning(md, p);;
+            } catch (CloneNotSupportedException cnse) {
+                // skip test if clone isn't supported
+                System.out.println("=> Clone not supported; skip!");
```
[the remark this answers is not in the document]
Comment (valeriepeng) [reply test/jdk/sun/security/pkcs11/MessageDigest/TestCloning.java:67]: Do you mean throw SkippedException if no digest algorithms are actually tested with Clone functionality testing? Existing NSS impl seems to support clone for non-SHA-3 digest impls.

File: test/jdk/sun/security/pkcs11/Signature/KeyAndParamCheckForPSS.java [hunk 9]
```diff
@@ -47,10 +47,7 @@ public static void main(String[] args) throws Exception {
 
     @Override
     public void main(Provider p) throws Exception {
-        Signature sig;
-        try {
-            sig = Signature.getInstance(SIGALG, p);
-        } catch (NoSuchAlgorithmException e) {
+        if (!PSSUtil.isSignatureSupported(p)) {
             System.out.println("Skip testing RSASSA-PSS" +
```
[the remark this answers is not in the document]
Comment (valeriepeng) [reply test/jdk/sun/security/pkcs11/Signature/KeyAndParamCheckForPSS.java:47]: Ok

File: test/jdk/sun/security/pkcs11/Signature/KeyAndParamCheckForPSS.java [hunk 10]
```diff
@@ -78,25 +75,20 @@ public void main(Provider p) throws Exception {
         runTest(p, 1040, "SHA3-512", "SHA3-512");
```
[the remark this answers is not in the document]
Comment (valeriepeng) [reply test/jdk/sun/security/pkcs11/Signature/KeyAndParamCheckForPSS.java:71]: Ok

File: test/jdk/sun/security/pkcs11/MessageDigest/TestCloning.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (rhalade) [reply test/jdk/sun/security/pkcs11/MessageDigest/TestCloning.java:67]: Yes, test should throw SkippedException when nothing is tested.

File: test/jdk/sun/security/pkcs11/Signature/KeyAndParamCheckForPSS.java [hunk 11]
```diff
@@ -108,58 +100,71 @@ private void runTest(Provider p, int keySize, String hashAlg,
...
             System.out.println("test#1: got expected IKE");
         }
 
+        // test#2 - setParameter then initVerify
+        sig = Signature.getInstance(SIGALG, p);
         sig.setParameter(paramsGood);
-        sig.initSign(priv);
-        System.out.println("test#1: pass");
+        sig.initVerify(pub);
+        System.out.println("test#2: good params pass");
```
[the remark this answers is not in the document]
Comment (valeriepeng) [reply test/jdk/sun/security/pkcs11/Signature/KeyAndParamCheckForPSS.java:125]: test#1 message matches with test#2 ones. Anyway, I removed the test output messages with your comments.

File: test/jdk/sun/security/pkcs11/Signature/SignatureTestPSS.java (hunk 2 above)
Comment (valeriepeng) [reply test/jdk/sun/security/pkcs11/Signature/SignatureTestPSS.java]: Ok. I initially thought that the SkippedException only applies to library not found and/or un-configured OS. With this extension of the meaning of SkippedException, many of the existing tests have to be updated to match.

File: test/jdk/sun/security/pkcs11/Signature/SignatureTestPSS.java (hunk 3 above)
Comment (valeriepeng) [reply test/jdk/sun/security/pkcs11/Signature/SignatureTestPSS.java:80]: This is just one digest algorithm. To mark the test skipped if none of the algorithms are supported will require different handling.

File: test/jdk/sun/security/pkcs11/PSSUtil.java (hunk 5 above)
Comment (valeriepeng) [reply test/jdk/sun/security/pkcs11/PSSUtil.java:34]: Removed.

File: test/jdk/sun/security/pkcs11/PSSUtil.java (hunk 4 above)
Comment (valeriepeng) [reply test/jdk/sun/security/pkcs11/PSSUtil.java:24]: Sure

File: test/jdk/sun/security/pkcs11/Signature/SignatureTestPSS.java (hunk 2 above)
Comment (rhalade) [reply test/jdk/sun/security/pkcs11/Signature/SignatureTestPSS.java]: Yes, I did #1 update with [JDK-8313206](https://bugs.openjdk.org/browse/JDK-8313206) and plan to do update more scenarios with [JDK-8313575](https://bugs.openjdk.org/browse/JDK-8313575)

File: test/jdk/sun/security/pkcs11/Signature/SignatureTestPSS.java (hunk 3 above)
Comment (rhalade) [reply test/jdk/sun/security/pkcs11/Signature/SignatureTestPSS.java:80]: Yes est should only throw `SkippedEXception` when all algos are skipped.

**Review by rhalade — CHANGES_REQUESTED** [review]

File: test/jdk/sun/security/pkcs11/PSSUtil.java (hunk 6 above)
Comment (valeriepeng) [reply test/jdk/sun/security/pkcs11/PSSUtil.java]: Ok.

**Comment by valeriepeng** [comment]
> Can you please also update https://github.com/openjdk/jdk/blob/master/test/jdk/sun/security/pkcs11/PKCS11Test.java#L1002 to use 3.91?

Are you suggesting to use 3.91 for all the OS configurations?
Based on my experiment, many of the mach5 linux x64 machines miss the required library for the NSS v3.91. I ended up using an older version of sqlite3 and test the changes locally. Thus, I have not changed the NSS version in the PKCS11Test.java.

**Comment by rhalade** [comment]
> 

Ok, lets address this update in a separate bug after validating all platforms.

**Comment by valeriepeng** [comment]
> > 
> 
> Ok, lets address this update in a separate bug after validating all platforms.

Sounds good to me.
Do you have any comment that I have yet address?

File: test/jdk/sun/security/pkcs11/MessageDigest/TestCloning.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (valeriepeng) [reply test/jdk/sun/security/pkcs11/MessageDigest/TestCloning.java:72]: I re-factored the test to throw SkippedException when no digest algorithm supports clone functionality.

File: test/jdk/sun/security/pkcs11/Signature/SignatureTestPSS.java (hunk 3 above)
Comment (valeriepeng) [reply test/jdk/sun/security/pkcs11/Signature/SignatureTestPSS.java:82]: Re-factored test to throw SkippedException if all algos are skipped.

File: test/jdk/sun/security/pkcs11/PSSUtil.java (hunk 5 above)
Comment (rhalade) [reply test/jdk/sun/security/pkcs11/PSSUtil.java:34]: DIGESTS is still not used.

**Review by rhalade — CHANGES_REQUESTED** [review]

File: test/jdk/sun/security/pkcs11/Signature/KeyAndParamCheckForPSS.java [hunk 12]
```diff
@@ -108,58 +103,69 @@ private void runTest(Provider p, int keySize, String hashAlg,
...
+            sig.setParameter(paramsGood);
+            sig.initSign(priv);
+            // algorithm support confirmed
+            skipTest = false;
+        } catch (Exception ex) {
+            if (s == PSSUtil.AlgoSupport.MAYBE) {
+                // confirmed to be unsupported; skip the rest of the test
+                System.out.println("=> Skip; no PSS support");
+                return;
+            }
```
[the remark this answers is not in the document]
Comment (valeriepeng) [reply test/jdk/sun/security/pkcs11/Signature/KeyAndParamCheckForPSS.java:117]: Sure, should not happen but good to check  to be sure.

File: test/jdk/sun/security/pkcs11/Signature/SignatureTestPSS.java [hunk 13]
```diff
@@ -56,88 +55,70 @@ public class SignatureTestPSS extends PKCS11Test {
...
-            MessageDigest md = MessageDigest.getInstance(hash, prov);
-            if (!hash.equalsIgnoreCase(mgfHash)) {
-                md = MessageDigest.getInstance(mgfHash, prov);
+            sig.setParameter(params);
+        } catch (InvalidAlgorithmParameterException iape) {
+            if (s == PSSUtil.AlgoSupport.MAYBE) {
+                // confirmed to be unsupported; skip the rest of the test
+                System.out.println("    => Skip; no PSS support");
+                return;
             }
```
[the remark this answers is not in the document]
Comment (valeriepeng) [reply test/jdk/sun/security/pkcs11/Signature/SignatureTestPSS.java:117]: Sure.

File: test/jdk/sun/security/pkcs11/PSSUtil.java (hunk 5 above)
Comment (valeriepeng) [reply test/jdk/sun/security/pkcs11/PSSUtil.java:34]: Missed to commit this file. Will update in the next commit.

**Review by rhalade — APPROVED** [review]

**Pull request closed without merging**